drivers/i2c/i2c_w91.c

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright (c) 2021-2023 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT telink_w91_i2c
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(i2c_telink);
+
+#include <zephyr/drivers/i2c.h>
+#include "i2c-priv.h"
+#include <zephyr/drivers/pinctrl.h>
+#include <ipc/ipc_based_driver.h>
+#include <stdlib.h>
+
+enum {
+	IPC_DISPATCHER_I2C_CONFIGURE_EVENT = IPC_DISPATCHER_I2C,
+	IPC_DISPATCHER_I2C_MASTER_READ_EVENT,
+	IPC_DISPATCHER_I2C_MASTER_WRITE_EVENT,
+};
+
+enum i2c_role {
+	I2C_ROLE_MASTER,
+	I2C_ROLE_SLAVE,
+};
+
+enum i2c_addr_len {
+	I2C_ADDR_LEN_7BIT,
+	I2C_ADDR_LEN_10BIT,
+};
+
+struct i2c_ipc_cfg {
+	enum i2c_role role;
+	enum i2c_addr_len addr_len;
+	uint32_t master_clock;
+	uint16_t slave_addr;
+	uint8_t dma_en;
+	uint8_t pull_up_en;
+};
+
+struct i2c_master_tx_req {
+	uint16_t addr;
+	uint32_t tx_len;
+	uint8_t *tx_buf;
+};
+
+struct i2c_master_rx_req {
+	uint16_t addr;
+	uint32_t rx_len;
+};
+
+struct i2c_master_rx_resp {
+	int err;
+	uint32_t len;
+	uint8_t *buffer;
+};
+
+/* I2C configuration structure */
+struct i2c_w91_cfg {
+	uint32_t bitrate;
+	const struct pinctrl_dev_config *pcfg;
+	uint8_t instance_id;
+};
+
+/* I2C data structure */
+struct i2c_w91_data {
+	struct k_sem mutex;
```

Comment:
Mutex should be mutex type...
Here I see that it's used as mutex so use mutex. There is a difference between them.